pkg/util/limiter/utilization.go

```diff
@@ -29,21 +30,32 @@ const (
 )
 
 type utilizationScanner interface {
-	// Scan returns CPU time in seconds and memory utilization in bytes, or an error.
+	// Scan returns CPU time in seconds and Go heap size in bytes, or an error.
 	Scan() (float64, uint64, error)
 }
 
-type procfsScanner struct {
+// combinedScanner scans /proc for CPU utilization and Go runtime for heap size.
+type combinedScanner struct {
 	proc procfs.Proc
 }
 
-func (s procfsScanner) Scan() (float64, uint64, error) {
+func (s combinedScanner) Scan() (float64, uint64, error) {
 	ps, err := s.proc.Stat()
 	if err != nil {
 		return 0, 0, errors.Wrap(err, "failed to get process stats")
 	}
 
-	return ps.CPUTime(), uint64(ps.ResidentMemory()), nil
+	var m runtime.MemStats
+	runtime.ReadMemStats(&m)
```

Comment:
this might end up being on the expensive side: causes some per-process caches to be flushed and stops all goroutines from executing.


```
func ReadMemStats(m *MemStats) {
	stopTheWorld("read mem stats")
...

	// Flush mcaches to mcentral before doing anything else.
	//
	// Flushing to the mcentral may in general cause stats to
	// change as mcentral data structures are manipulated.
	systemstack(flushallmcaches)
}
```

Have you ran any benchmarks to see if this ends up affecting performance if we run it every second? Perhaps something on `utilizationScanner` itself or something more general like the ingester push and query benchmarks.

Reply:
Wow, I had no idea this could actually affect perf.

Reply:
@dimitarvdimitrov in a piece of seemingly good news, I found a Go [issue](https://github.com/golang/go/issues/13613) touching on `runtime.ReadMemStats` performance implications, and found that it was closed by a [commit](https://github.com/golang/go/commit/4a7cf960c38d72e9f0c6f00e46e013be2a35d56e) from 2017 that supposedly makes it "stop the world" for at most 25µs.

Reply:
I added a benchmark for `combinedScanner`. I find that calling `combinedScanner.Scan` only takes ~15 microseconds more than `procfsScanner` in the main branch (which `combinedScanner` replaces in this PR). That should be acceptable, and is in line with the claimed upper bound of `runtime.ReadMemStats` (25 microseconds).

Reply:
thanks for finding this out